test/test_async.py

```diff
@@ -0,0 +1,160 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+#
+# Copyright (c) 2012-2020 Snowflake Computing Inc. All right reserved.
+#
+import time
+
+import pytest
+
+from snowflake.connector import ProgrammingError
+from snowflake.connector.constants import QueryStatus
+from snowflake.connector.errors import InterfaceError
+
+# Mark all tests in this file to time out after 2 minutes to prevent hanging forever
+pytestmark = pytest.mark.timeout(120)
+
+
+def test_async_exec(conn_cnx):
+    """Tests whether simple async query execution works.
+
+    Runs a query that takes a few seconds to finish and then totally closes connection
+    to Snowflake. Then waits enough time for that query to finish, opens a new connection
+    and fetches results. It also tests QueryStatus related functionality too.
+    """
+    with conn_cnx() as con:
+        with con.cursor() as cur:
+            cur.execute('select count(*) from table(generator(timeLimit => 5))', exec_async=True)
+            q_id = cur.sfqid
+            status = con.get_query_status(q_id)
+            assert status in (QueryStatus.RUNNING, QueryStatus.NO_DATA)
+            assert con.is_still_running(status)
+    time.sleep(10)
+    with conn_cnx() as con:
+        with con.cursor() as cur:
+            status = con.get_query_status(q_id)
+            assert status == QueryStatus.SUCCESS
+            cur.get_results_from_sfqid(q_id)
+            assert len(cur.fetchall()) == 1
+
+
+def test_async_error(conn_cnx):
+    """Tests whether simple async query error retrieval works.
+
+    Runs a query that will fail to execute and then tests that if we tried to get results for the query
+    then that would raise an exception. It also tests QueryStatus related functionality too.
+    """
+    with conn_cnx() as con:
+        with con.cursor() as cur:
+            cur.execute('select * from nonexistentTable', exec_async=True)
+            q_id = cur.sfqid
+            while con.is_still_running(con.get_query_status(q_id)):
+                time.sleep(1)
+            status = con.get_query_status(q_id)
+            assert status == QueryStatus.FAILED_WITH_ERROR
+            assert con.is_an_error(status)
+            with pytest.raises(ProgrammingError):
+                con.get_query_status_throw_if_error(q_id)
+            with pytest.raises(ProgrammingError):
+                cur.get_results_from_sfqid(q_id)
+
+
+def test_mix_sync_async(conn_cnx):
```

Comment:
A small question, what will happen if we start a sync query while there is already a running async query? I guess that's supported? If yes, then do we want to add a simple test for it?

Reply:
One step ahead of you. Line 73 and 80 do that here.

Reply:
Sorry, actually I mean to run a **sync** query when there is a still-running async query. But nvm, it looks like not an error-prone scenario to me now. I think we are good 😁